File: tests/web/test_idm.mjs
// Created by Matthew Valancy
// Copyright 2026 Valpatel Software LLC
// Licensed under AGPL-3.0 — see LICENSE for details.
// Tests for tritium-lib/web/sim/idm.js — Intelligent Driver Model
// Run: node tests/web/test_idm.mjs

import { idmAcceleration, idmFreeFlow, idmStep, IDM_DEFAULTS, ROAD_SPEEDS } from '../../web/sim/idm.js';
import assert from 'node:assert';

let pass = 0, fail = 0;
function test(name, fn) {
    try { fn(); pass++; console.log(`  PASS: ${name}`); }
    catch (e) { fail++; console.log(`  FAIL: ${name} — ${e.message}`); }
}

console.log('=== IDM Tests ===');

// ── IDM_DEFAULTS ──────────────────────────────────────────────────

test('IDM_DEFAULTS has all required fields', () => {
    assert.strictEqual(typeof IDM_DEFAULTS.v0, 'number');
    assert.strictEqual(typeof IDM_DEFAULTS.a, 'number');
    assert.strictEqual(typeof IDM_DEFAULTS.b, 'number');
    assert.strictEqual(typeof IDM_DEFAULTS.s0, 'number');
    assert.strictEqual(typeof IDM_DEFAULTS.T, 'number');
    assert.strictEqual(typeof IDM_DEFAULTS.delta, 'number');
    assert.strictEqual(IDM_DEFAULTS.delta, 4);
});

test('ROAD_SPEEDS has expected road classes', () => {
    assert.strictEqual(typeof ROAD_SPEEDS.motorway, 'number');
    assert.strictEqual(typeof ROAD_SPEEDS.residential, 'number');
    assert.strictEqual(typeof ROAD_SPEEDS.service, 'number');
    assert.ok(ROAD_SPEEDS.motorway > ROAD_SPEEDS.residential, 'motorway faster than residential');
    assert.ok(ROAD_SPEEDS.residential > ROAD_SPEEDS.service, 'residential faster than service');
});

// ── idmFreeFlow ───────────────────────────────────────────────────

test('idmFreeFlow: zero speed gives max acceleration', () => {
    const acc = idmFreeFlow(0, IDM_DEFAULTS);
    // At v=0: a * (1 - 0) = a
    assert.ok(Math.abs(acc - IDM_DEFAULTS.a) < 0.001, `expected ~${IDM_DEFAULTS.a}, got ${acc}`);
});

test('idmFreeFlow: at desired speed, acceleration is zero', () => {
    const acc = idmFreeFlow(IDM_DEFAULTS.v0, IDM_DEFAULTS);
    // At v=v0: a * (1 - 1) = 0
    assert.ok(Math.abs(acc) < 0.001, `expected ~0, got ${acc}`);
});

test('idmFreeFlow: above desired speed, acceleration is negative', () => {
    const acc = idmFreeFlow(IDM_DEFAULTS.v0 * 1.5, IDM_DEFAULTS);
    assert.ok(acc < 0, `expected negative, got ${acc}`);
});

test('idmFreeFlow: half desired speed gives positive acceleration', () => {
    const acc = idmFreeFlow(IDM_DEFAULTS.v0 / 2, IDM_DEFAULTS);
    assert.ok(acc > 0, `expected positive, got ${acc}`);
    assert.ok(acc < IDM_DEFAULTS.a, 'should be less than max acceleration');
});

// ── idmAcceleration ───────────────────────────────────────────────

test('idmAcceleration: large gap and same speed behaves like free flow', () => {
    const freeAcc = idmFreeFlow(5, IDM_DEFAULTS);
    const acc = idmAcceleration(5, 1000, 5, IDM_DEFAULTS);
    // With a 1000m gap and no speed difference, interaction term is negligible
    assert.ok(Math.abs(acc - freeAcc) < 0.1, `expected ~${freeAcc}, got ${acc}`);
});

test('idmAcceleration: very small gap causes heavy braking', () => {
    const acc = idmAcceleration(10, 1.0, 0, IDM_DEFAULTS);
    assert.ok(acc < -2.0, `expected strong braking, got ${acc}`);
});

test('idmAcceleration: clamped to minimum -9 m/s^2', () => {
    const acc = idmAcceleration(20, 0.5, 0, IDM_DEFAULTS);
    assert.ok(acc >= -9.0, `expected >= -9, got ${acc}`);
});

test('idmAcceleration: approaching faster than leader causes braking', () => {
    const acc = idmAcceleration(15, 10, 5, IDM_DEFAULTS);
    // Big speed difference (15 vs 5) at close range = braking
    assert.ok(acc < 0, `expected negative (braking), got ${acc}`);
});

test('idmAcceleration: leader faster than ego reduces desired gap', () => {
    // When leader is faster, dv is negative, which reduces s*
    const accFaster = idmAcceleration(5, 10, 10, IDM_DEFAULTS);
    const accSame = idmAcceleration(5, 10, 5, IDM_DEFAULTS);
    assert.ok(accFaster >= accSame, 'faster leader should mean less braking');
});

test('idmAcceleration: gap cannot be zero (clamped to 0.5)', () => {
    // Should not throw or produce NaN/Infinity
    const acc = idmAcceleration(5, 0, 5, IDM_DEFAULTS);
    assert.ok(Number.isFinite(acc), `expected finite, got ${acc}`);
});

// ── idmStep ───────────────────────────────────────────────────────

test('idmStep: positive acceleration increases speed', () => {
    const { v, ds } = idmStep(5, 1.0, 1.0);
    assert.strictEqual(v, 6);
    assert.ok(ds > 5, 'distance should be more than v*dt due to acceleration');
});

test('idmStep: speed does not go below zero', () => {
    const { v, ds } = idmStep(1, -5, 1.0);
    assert.strictEqual(v, 0);
    assert.ok(ds >= 0, 'distance should be non-negative');
});

test('idmStep: zero acceleration, constant speed', () => {
    const { v, ds } = idmStep(10, 0, 0.5);
    assert.strictEqual(v, 10);
    assert.ok(Math.abs(ds - 5.0) < 0.001, `expected 5.0, got ${ds}`);
});

test('idmStep: small timestep', () => {
    const { v, ds } = idmStep(10, 1.0, 0.01);
    assert.ok(Math.abs(v - 10.01) < 0.001, `expected ~10.01, got ${v}`);
    assert.ok(ds > 0, 'should move forward');
});

test('idmStep: distance uses kinematic equation', () => {
    // ds = v*dt + 0.5*a*dt^2 = 10*1 + 0.5*2*1 = 11
    const { ds } = idmStep(10, 2, 1.0);
    assert.ok(Math.abs(ds - 11) < 0.001, `expected 11, got ${ds}`);
});

console.log(`\n=== Results: ${pass} passed, ${fail} failed ===`);
process.exit(fail > 0 ? 1 : 0);

File: tests/sim_engine/test_idm.py
# Created by Matthew Valancy
# Copyright 2026 Valpatel Software LLC
# Licensed under AGPL-3.0 — see LICENSE for details.
"""Tests for tritium_lib.sim_engine.idm -- Intelligent Driver Model."""

import math

import pytest

from tritium_lib.sim_engine.idm import (
    IDMParams,
    IDM_DEFAULTS,
    IDMStepResult,
    ROAD_SPEEDS,
    VEHICLE_IDM_PROFILES,
    idm_acceleration,
    idm_free_flow,
    idm_step,
    get_idm_for_road,
)


# ===================================================================
# IDMParams
# ===================================================================


class TestIDMParams:
    def test_defaults(self):
        p = IDMParams()
        assert p.v0 == 12.0
        assert p.a == 1.4
        assert p.b == 2.0
        assert p.s0 == 2.0
        assert p.T == 1.5
        assert p.delta == 4

    def test_custom(self):
        p = IDMParams(v0=30.0, a=2.0, b=3.0, s0=3.0, T=1.0, delta=6)
        assert p.v0 == 30.0
        assert p.a == 2.0
        assert p.delta == 6

    def test_idm_defaults_singleton(self):
        assert IDM_DEFAULTS.v0 == 12.0
        assert IDM_DEFAULTS.T == 1.5


# ===================================================================
# idm_acceleration
# ===================================================================


class TestIDMAcceleration:
    def test_stopped_no_leader(self):
        """Stopped car with no leader should accelerate positively."""
        acc = idm_acceleration(0.0, 1000.0, 0.0)
        assert acc > 0.0

    def test_at_desired_speed_no_leader(self):
        """At desired speed with large gap, acceleration should be near zero."""
        p = IDMParams(v0=12.0)
        acc = idm_acceleration(12.0, 1000.0, 12.0, p)
        assert abs(acc) < 0.5  # nearly zero

    def test_close_to_stopped_leader(self):
        """Close to a stopped leader, should brake hard."""
        acc = idm_acceleration(10.0, 3.0, 0.0)
        assert acc < -1.0  # braking

    def test_large_gap_accelerates(self):
        """Large gap with slow speed should give positive acceleration."""
        acc = idm_acceleration(5.0, 200.0, 15.0)
        assert acc > 0.0

    def test_approaching_leader_brakes(self):
        """Going faster than leader at small gap should brake."""
        acc = idm_acceleration(15.0, 10.0, 5.0)
        assert acc < 0.0

    def test_same_speed_comfortable_gap(self):
        """Same speed with comfortable gap should have mild acceleration."""
        p = IDMParams(v0=12.0, T=1.5, s0=2.0)
        # Gap = v0 * T + s0 = 12 * 1.5 + 2 = 20m -- comfortable distance
        acc = idm_acceleration(10.0, 20.0, 10.0, p)
        # Should be slightly positive (approaching desired speed, comfortable gap)
        assert acc > -1.0

    def test_braking_limit(self):
        """Acceleration should be clamped to -9 m/s^2 (1g)."""
        acc = idm_acceleration(20.0, 0.5, 0.0)
        assert acc >= -9.0

    def test_acceleration_limit(self):
        """Acceleration should not exceed max acceleration."""
        p = IDMParams(a=1.4)
        acc = idm_acceleration(0.0, 1000.0, 0.0, p)
        assert acc <= p.a

    def test_zero_v0(self):
        """v0=0 should not crash (edge case)."""
        p = IDMParams(v0=0.0)
        acc = idm_acceleration(5.0, 10.0, 5.0, p)
        assert isinstance(acc, float)

    def test_negative_gap_clamped(self):
        """Negative or very small gap should be clamped, not cause division by zero."""
        acc = idm_acceleration(5.0, 0.01, 0.0)
        assert isinstance(acc, float)
        assert not math.isnan(acc)
        assert not math.isinf(acc)


# ===================================================================
# idm_free_flow
# ===================================================================


class TestIDMFreeFlow:
    def test_stopped_accelerates(self):
        """Stopped vehicle in free flow should get max acceleration."""
        acc = idm_free_flow(0.0)
        assert acc == pytest.approx(IDM_DEFAULTS.a, abs=0.01)

    def test_at_desired_speed(self):
        """At desired speed, free-flow acceleration should be zero."""
        p = IDMParams(v0=12.0)
        acc = idm_free_flow(12.0, p)
        assert acc == pytest.approx(0.0, abs=0.01)

    def test_above_desired_speed(self):
        """Above desired speed, should decelerate."""
        p = IDMParams(v0=10.0)
        acc = idm_free_flow(15.0, p)
        assert acc < 0.0

    def test_half_speed(self):
        """At half desired speed, should have significant acceleration."""
        p = IDMParams(v0=12.0, a=1.4, delta=4)
        acc = idm_free_flow(6.0, p)
        # (1 - (6/12)^4) = (1 - 0.0625) = 0.9375
        expected = 1.4 * 0.9375
        assert acc == pytest.approx(expected, abs=0.01)

    def test_zero_v0(self):
        p = IDMParams(v0=0.0)
        acc = idm_free_flow(5.0, p)
        assert acc == 0.0


# ===================================================================
# idm_step
# ===================================================================


class TestIDMStep:
    def test_positive_acceleration(self):
        result = idm_step(5.0, 1.0, 0.1)
        assert isinstance(result, IDMStepResult)
        assert result.v == pytest.approx(5.1, abs=0.01)
        assert result.ds > 0

    def test_braking(self):
        result = idm_step(10.0, -3.0, 0.1)
        assert result.v == pytest.approx(9.7, abs=0.01)
        assert result.ds > 0

    def test_hard_braking_clamps_speed(self):
        """Speed should not go negative."""
        result = idm_step(1.0, -20.0, 1.0)
        assert result.v == 0.0

    def test_zero_speed_zero_acc(self):
        result = idm_step(0.0, 0.0, 0.1)
        assert result.v == 0.0
        assert result.ds == 0.0

    def test_distance_formula(self):
        """ds = v*dt + 0.5*a*dt^2"""
        v, a, dt = 10.0, 2.0, 0.1
        result = idm_step(v, a, dt)
        expected_ds = v * dt + 0.5 * a * dt * dt
        assert result.ds == pytest.approx(expected_ds, abs=0.001)


# ===================================================================
# ROAD_SPEEDS
# ===================================================================


class TestRoadSpeeds:
    def test_motorway_fastest(self):
        assert ROAD_SPEEDS["motorway"] >= ROAD_SPEEDS["residential"]

    def test_service_slowest(self):
        assert ROAD_SPEEDS["service"] <= ROAD_SPEEDS["residential"]

    def test_all_positive(self):
        for cls, speed in ROAD_SPEEDS.items():
            assert speed > 0, f"{cls} speed must be positive"


# ===================================================================
# VEHICLE_IDM_PROFILES
# ===================================================================


class TestVehicleProfiles:
    def test_sedan_exists(self):
        assert "sedan" in VEHICLE_IDM_PROFILES

    def test_truck_slower(self):
        assert VEHICLE_IDM_PROFILES["truck"].v0 < VEHICLE_IDM_PROFILES["sedan"].v0

    def test_motorcycle_faster(self):
        assert VEHICLE_IDM_PROFILES["motorcycle"].v0 > VEHICLE_IDM_PROFILES["sedan"].v0

    def test_all_profiles_valid(self):
        for name, params in VEHICLE_IDM_PROFILES.items():
            assert params.v0 > 0, f"{name} v0"
            assert params.a > 0, f"{name} a"
            assert params.b > 0, f"{name} b"
            assert params.s0 > 0, f"{name} s0"
            assert params.T > 0, f"{name} T"


# ===================================================================
# get_idm_for_road
# ===================================================================


class TestGetIDMForRoad:
    def test_residential(self):
        p = get_idm_for_road("residential", speed_variation=0.0)
        assert p.v0 == pytest.approx(ROAD_SPEEDS["residential"], abs=0.01)

    def test_motorway(self):
        p = get_idm_for_road("motorway", speed_variation=0.0)
        assert p.v0 == pytest.approx(ROAD_SPEEDS["motorway"], abs=0.01)

    def test_unknown_road_class_fallback(self):
        p = get_idm_for_road("alien_highway", speed_variation=0.0)
        assert p.v0 == pytest.approx(10.0, abs=0.01)  # default fallback

    def test_variation_within_range(self):
        """With variation, speed should be within +/-10% of base."""
        base = ROAD_SPEEDS["residential"]
        for _ in range(20):
            p = get_idm_for_road("residential", speed_variation=0.1)
            assert base * 0.85 <= p.v0 <= base * 1.15

    def test_preserves_base_params(self):
        base = IDMParams(a=2.0, b=3.0, s0=4.0, T=1.0)
        p = get_idm_for_road("residential", base_params=base, speed_variation=0.0)
        assert p.a == 2.0
        assert p.b == 3.0
        assert p.s0 == 4.0
        assert p.T == 1.0


# ===================================================================
# IDM behavioral properties (higher-level integration)
# ===================================================================


class TestIDMBehavior:
    """Test that IDM produces physically reasonable traffic behavior."""

    def test_equilibrium_speed(self):
        """A single car should converge toward v0 after many steps."""
        p = IDMParams(v0=12.0, a=1.4)
        v = 0.0
        dt = 0.1
        for _ in range(500):
            acc = idm_free_flow(v, p)
            result = idm_step(v, acc, dt)
            v = result.v
        assert v == pytest.approx(12.0, abs=0.5)

    def test_platoon_stabilizes(self):
        """Two cars following each other should stabilize at a safe gap."""
        p = IDMParams(v0=12.0, s0=2.0, T=1.5)
        leader_v = 10.0  # constant speed leader
        follower_v = 0.0
        gap = 50.0
        dt = 0.1

        for _ in range(1000):
            acc = idm_acceleration(follower_v, gap, leader_v, p)
            result = idm_step(follower_v, acc, dt)
            gap -= (result.ds - leader_v * dt)
            follower_v = result.v

        # Follower should match leader speed
        assert follower_v == pytest.approx(leader_v, abs=1.0)
        # Gap should be > s0 (minimum gap)
        assert gap > p.s0

    def test_emergency_stop(self):
        """Car approaching a stopped obstacle should stop before collision."""
        p = IDMParams(v0=12.0, s0=2.0, T=1.5, b=2.0)
        v = 12.0
        gap = 30.0
        dt = 0.05

        for _ in range(2000):
            acc = idm_acceleration(v, gap, 0.0, p)
            result = idm_step(v, acc, dt)
            gap -= result.ds
            v = result.v
            if v < 0.01:
                break

        # Car should stop before collision
        assert gap > 0, f"Car collided! Remaining gap: {gap:.2f}"
        assert v < 0.1, f"Car didn't stop: v={v:.2f}"
